<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.0 Transitional//EN">
<HTML>
<HEAD>
	<META HTTP-EQUIV="CONTENT-TYPE" CONTENT="text/html; charset=utf-8">
	<TITLE></TITLE>
	<META NAME="GENERATOR" CONTENT="OpenOffice.org 2.3  (Linux)">
	<META NAME="CREATED" CONTENT="20081210;10343900">
	<META NAME="CHANGED" CONTENT="20081210;15062400">
	<STYLE TYPE="text/css">
	<!--
		@page { size: 8.5in 11in; margin: 0.79in }
		P { margin-bottom: 0.08in }
		H3 { margin-bottom: 0.08in }
		H3.western { font-family: "Liberation Sans", sans-serif }
	-->
	</STYLE>
</HEAD>
<BODY LANG="en-US" DIR="LTR">
<H3 CLASS="western">Search Option: Gene</H3>
<P STYLE="margin-bottom: 0in">This is not actually a search option,
you can only input a single gene name here. Databases with gene names
that are well curated are available for a few species only (e.g.
Human and mouse). Fansnip incorporates the gene database from NCBI,
which in turn uses curated databases if available (e.g. Hugo for
human gene names). For species without curated databases NCBI will
assign gene names based on homologue predictions.</P>
<P STYLE="margin-bottom: 0in">	Fansnip will only retrieve a gene name
if it exists in the NCBI database. Alias databases have not been
implemented as dealing with ambiguities often can only be done by
user intervention. The rationale behind Fansnip is to minimize user
intervention. There is one exception: if the gene name itself is not
present for a non-human species, but it is the same as a human gene
name, the orthologue of that gene will be retrieve on the premise
that gene names will only be identical if there is some kind of
homology relationship. This may not always be the case.</P>
<P STYLE="margin-bottom: 0in">	As Fansnip uses primarily Ensembl gene
id's no gene can be retrieved if there is not a proper cross database
(NCBI-EnsEMBL) annotation. Unfortunately this is the case for a
sizeable portion of the genes.</P>
<P STYLE="margin-bottom: 0in"><BR>
</P>
<H3 CLASS="western">What to expect:</H3>
<P STYLE="margin-bottom: 0in">Let's assume you have typed in a valid
gene name, say 'IGF1'.</P>
<P STYLE="margin-bottom: 0in">It may take some time for results to
appear. Then you will see:</P>
<P STYLE="margin-bottom: 0in"><STRONG>chicken gene </STRONG><A HREF="http://www.ncbi.nlm.nih.gov/sites/entrez?db=gene&amp;cmd=search&amp;term=418090" TARGET="_blank"><STRONG><EM>IGF1</EM></STRONG><STRONG>
</STRONG></A><STRONG>(insulin-like growth factor 1 (somatomedin C))
has NCBI gene id </STRONG><STRONG><EM>418090 </EM></STRONG><STRONG>and
EnsEMBL gene id </STRONG><STRONG><EM><A HREF="http://www.ensembl.org/Gallus_gallus/geneview?gene=ENSGALG00000012755" TARGET="_blank">ENSGALG00000012755</A></EM></STRONG></P>
<P STYLE="margin-bottom: 0in">You will also notice a button stating
'submit'. If you press this button, the gene(s) that were found can
be process in the SNPsearch phase.</P>
<P STYLE="margin-bottom: 0in"><BR>
</P>
<P STYLE="margin-bottom: 0in">If you did not submit a valid name you
will be notified about it and you can choose to start a new query by
pressing 'new'.</P>
</BODY>
</HTML>